cmd/thanos/endpointset.go

```diff
@@ -0,0 +1,374 @@
+// Copyright (c) The Thanos Authors.
+// Licensed under the Apache License 2.0.
+
+package main
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/oklog/run"
+	"google.golang.org/grpc"
+	"gopkg.in/yaml.v3"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/discovery"
+	"github.com/prometheus/prometheus/discovery/file"
+	"github.com/prometheus/prometheus/discovery/targetgroup"
+
+	"github.com/thanos-io/thanos/pkg/component"
+	"github.com/thanos-io/thanos/pkg/discovery/cache"
+	"github.com/thanos-io/thanos/pkg/discovery/dns"
+	"github.com/thanos-io/thanos/pkg/errors"
+	"github.com/thanos-io/thanos/pkg/extgrpc"
+	"github.com/thanos-io/thanos/pkg/extkingpin"
+	"github.com/thanos-io/thanos/pkg/extprom"
+	"github.com/thanos-io/thanos/pkg/query"
+	"github.com/thanos-io/thanos/pkg/runutil"
+)
+
+// fileContent is the interface of methods that we need from extkingpin.PathOrContent.
+// We need to abstract it for now so we can implement a default if the user does not provide one.
+type fileContent interface {
+	Content() ([]byte, error)
+	Path() string
+}
+
+// emptyPathContent is a "fileContent" implementation to use when nothing was provided.
+// we need this for some convenience while it is possible to not provide a `--endpoint.sd-config`
+// flag and just use the static `--endpoint` flags.
+// Once we decided to make this flag required and deprecate the "--endpoint" families of flags
+// this can be discarded again.
+type emptyPathContent struct {
+}
+
+func (t *emptyPathContent) Content() ([]byte, error) {
```

Comment:
receive/limiter.go has `type nopConfigContent struct{}`. Should we export this & reuse it?